Fix predicate pushdown for outer joins

james727 opened the pull request:
# Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/1586

# What changes are included in this PR?
This change does a couple of things:
1. Does not attempt to push down any predicates for the nullable (non-preserved) side of a given join
2. Bails on the optimization to duplicate post-join filters on a column involved in the join equality to the other side of the join, if the other side of the join is nullable.

I'll elaborate on point 2 here as I think the above is a bit difficult to parse. Consider the following query:
```sql
SELECT *
FROM t1 JOIN t2 
  on t1.id = t2.uid 
WHERE t1.id > 1
```
Right now this is being rewritten as:
```sql
SELECT *
FROM t1 JOIN t2 
  on t1.id = t2.uid 
WHERE t1.id > 1 AND t2.uid > 1 -- Duplicate filter to t2.uid, so it can be pushed down
```
This is correct in the case of certain joins. However, I believe that in cases where the other join side is non-preserved, we should not apply this optimization. For example, if we change the above join to a `LEFT` join, the query rewrite becomes:
```sql
SELECT *
FROM t1 LEFT JOIN t2 
  on t1.id = t2.uid 
WHERE t1.id > 1 AND t2.uid > 1
```
The additional filter on `t2.uid` is wrong - as it omits null rows that would have been present in the original query.

xudong963 commented:
Oh, thanks @james727 ! -- I'll review it later.

alamb commented:
I also hope to review this tomorrow

liukun4515 commented:
> # Which issue does this PR close?
> Closes #1586
> 
> # What changes are included in this PR?
> This change does a couple of things:
> 
> 1. Does not attempt to push down any predicates for the nullable (non-preserved) side of a given join
> 2. Bails on the optimization to duplicate post-join filters on a column involved in the join equality to the other side of the join, if the other side of the join is nullable.
> 
> I'll elaborate on point 2 here as I think the above is a bit difficult to parse. Consider the following query:
> 
> ```sql
> SELECT *
> FROM t1 JOIN t2 
>   on t1.id = t2.uid 
> WHERE t1.id > 1
> ```
> 
> Right now this is being rewritten as:
> 
> ```sql
> SELECT *
> FROM t1 JOIN t2 
>   on t1.id = t2.uid 
> WHERE t1.id > 1 AND t2.uid > 1 -- Duplicate filter to t2.uid, so it can be pushed down
> ```
> 
> This is correct in the case of certain joins. However, I believe that in cases where the other join side is non-preserved, we should not apply this optimization. For example, if we change the above join to a `LEFT` join, the query rewrite becomes:
> 
> ```sql
> SELECT *
> FROM t1 LEFT JOIN t2 
>   on t1.id = t2.uid 
> WHERE t1.id > 1 AND t2.uid > 1
> ```
> 
> The additional filter on `t2.uid` is wrong - as it omits null rows that would have been present in the original query.

great fix

liukun4515 commented:
great fix.

alamb commented on datafusion/tests/sql/joins.rs:
```diff
@@ -210,6 +210,26 @@ async fn left_join_unbalanced() -> Result<()> {
+async fn left_join_filter_pushdown() -> Result<()> {
+    // Since t2 is the non-preserved side of the join, we cannot push down a NULL filter.
+    let mut ctx = create_join_context_with_nulls()?;
+    let sql = "SELECT t1_id, t2_id, t2_name FROM t1 LEFT JOIN t2 ON t1_id = t2_id WHERE t2_name IS NULL ORDER BY t1_id";
```
👍  this does match postgres:

```sql
alamb=# create table t1(t1_id, t1_name) as select * from (values (11, 'a'), (22, 'b'), (33, 'c'), (44, 'd'), (77, 'e')) as sq;
SELECT 5
alamb=# create table t2(t2_id, t2_name) as select * from (values (11, 'z'), (22, null), (44, 'x'), (55, 'w')) as sq;
SELECT 4
alamb=# SELECT t1_id, t2_id, t2_name FROM t1 LEFT JOIN t2 ON t1_id = t2_id WHERE t2_name IS NULL ORDER BY t1_id;
 t1_id | t2_id | t2_name 
-------+-------+---------
    22 |    22 | 
    33 |       | 
    77 |       | 
(3 rows)

```


I think the key would be to test adding a filter on `t2_id` however (as that is the predicate type that is created)

alamb commented on datafusion/src/optimizer/filter_push_down.rs:
```diff
@@ -253,33 +176,107 @@ fn split_members<'a>(predicate: &'a Expr, predicates: &mut Vec<&'a Expr>) {
+            JoinType::Inner => (true, true),
+            JoinType::Left => (true, false),
+            JoinType::Right => (false, true),
+            JoinType::Full => (false, false),
```
Both sides are preserved in a ` FULL` join (as in this should probably be (`true`, `true`)

james727 replied:
Ah - my impression was that `FULL` was a `FULL OUTER` join - i.e. both sides are nullable/non-preserved? Is this not right? I might have mixed it up with `CrossJoin`.

alamb replied:
`Full` is `FULL OUTER` join... I think in that case the inputs are both preserved (as in you can't push filters down through the join if it isn't for that column)

Though that doesn't seem consistent with `JoinType::Inner`

Maybe 
```rust
            JoinType::Inner => (false, false),
``` 
as predicates *can* be pushed down both sides there 🤔

james727 replied:
Maybe whether a side is `nullable` is the better distinction here? I see your point about both sides of the `FULL` join being preserved, but you still can't push filters because they can provide extra null rows.

alamb replied:
I think there are two different things making this confusing:

1.  for predicates that are `null preserving` (like `t1 <= 5`) in the `WHERE` clause it is actually fine to create a predicate like `t2 < 5` even for outer joins (try it with postgres, you'll find that predicate filters out any rows that didn't have matches on either side so it is ok)
2. For predicates that are not `null preserving` (like `t1 is NULL`) you *CAN'T* push predicates down (which I think is the core of this PR)

To make things even more confusing, the rules of when you can push / not push are different if the predicate is in the `ON` clause)

Also I don't think DataFusion currently tracks the `null preserving` property for `Exprs` -- so the safe (correct thing) is to treat all predicates like they may not be `null preserving` and not create/push them for outer joins. A follow on exercise would be to allow more predicates to be pushed down by checking `null preserving`

alamb commented on datafusion/src/optimizer/filter_push_down.rs:
```diff
@@ -253,33 +176,107 @@ fn split_members<'a>(predicate: &'a Expr, predicates: &mut Vec<&'a Expr>) {
+    schema: &DFSchema,
+    preserved: bool,
+) -> Predicates<'a> {
+    if !preserved {
```
I think this check is wrong -- you can't push predicates to the *preserved* side

```suggestion
    if preserved {
```

james727 replied:
Oh jeez - facepalm

james727 replied:
I thought the idea was that for the nullable side of the join, you cannot push down filters (but you may be able to rewrite the join to an `INNER` and _then_ push down filters), but for the preserved/non-nullable side you can. This was my takeaway from https://github.com/apache/arrow-datafusion/issues/1586 and from reading around online.

E.g. - it seems safe to push down the filter in this query:
```
SELECT *
FROM t1 LEFT JOIN t2
ON t1.x = t2.y
WHERE t1.z > 5
```
While if the filter was on a column of `t2`, we could not push it down.

There is a good chance I have this totally wrong but wanted to double check before proceeding with the suggested change.

alamb reviewed (commented):
First of all @james727  thank you for looking into this -- NULL semantics are some of the trickiest parts of SQL and I salute you and @xudong963  for working on this.  ❤️ 

I *think* there are still some subtle bugs here (though I admit it is tricky) as I have highlighted; The code is really nice and I think it is very close. 

I think we can add some additional tests based on the examples in https://github.com/apache/arrow-datafusion/issues/1321 (expected postgres behavior) as well as in https://github.com/apache/arrow-datafusion/pull/1339#issuecomment-974191228 and https://github.com/apache/arrow-datafusion/pull/1339#discussion_r753455014. I recommend we port those tests to datafusion as part of this PR.

james727 commented on datafusion/src/optimizer/filter_push_down.rs:
```diff
@@ -1139,6 +1136,172 @@ mod tests {
+        let expected = "\
+        Filter: #test2.a <= Int64(1)\
+        \n  Join: Using #test.a = #test2.a\
+        \n    Filter: #test.a <= Int64(1)\
```
Oh man, great catch! I think this gets back to some previous discussion around filters that preserve nulls vs filters that do not preserve nulls. It seems this filter duplication logic is valid for filters that do not preserve nulls, but breaks when the filter preserves nulls (as shown above).

E.g. - correct me if I'm wrong I  _think_ that duplicating the filter is correct in the following query:
```
SELECT *
FROM t1 LEFT JOIN t2
ON t1.x = t2.y
WHERE t2.y > 5 -- Can duplicate filter for t1.x > 5
```

What do you think about going back to the logic in [this commit](https://github.com/apache/arrow-datafusion/pull/1618/commits/c80772b5ce2c1a6419cefb247d55ec9ef2ef9dbd) that just skips this optimization for non-inner joins (before I got fancy with it). If a filter does not preserve nulls, we will be able to rewrite the join as an `INNER` join anyways, so once that optimization is implemented we'll get the correct duplication logic with another pass of this optimizer.

james727 commented:
Thanks for the review @alamb! We already have some active discussion on this PR but I'd propose the following next steps on my end:
1. Add the suggested tests (and fix any associated breakage)
2. Roll back the filter duplication optimization to only apply for `INNER` joins (as discussed in https://github.com/apache/arrow-datafusion/pull/1618#discussion_r789951842)

I think the other main open thread is around the term `preserved` and its definition, it seems there is some misalignment. What are your thoughts on the best way to decide on the terminology/definitions to use here (and document it going forward)?

alamb commented:
> Roll back the filter duplication optimization to only apply for INNER joins (as discussed in Fix predicate pushdown for outer joins #1618 (comment))

I think this is a good (safe) outcome. Pushing predicates down through outer joins can be treated as a follow on, more sophisticated optimization 👍

james727 commented on datafusion/tests/sql/joins.rs:
```diff
@@ -210,6 +210,26 @@ async fn left_join_unbalanced() -> Result<()> {
 
+#[tokio::test]
+async fn left_join_filter_pushdown() -> Result<()> {
+    // Since t2 is the non-preserved side of the join, we cannot push down a NULL filter.
```
I think in this case, we can't even rewrite the join, since `IS NULL` by definition does not remove nulls. So this should still be valid? I'll try and clarify a bit in the comment though.

james727 commented:
@alamb I think this is ready for another look. I've updated the logic as discussed (do not duplicate filters on the join column unless it's an inner join) and added a bunch of tests. I also extended the comment on `lr_is_preserved` to hopefully be clear about how I've defined "preserved" in this PR. I think the way I have defined it, the logic is correct, but I'm curious your thoughts on whether this is the best concept to move forward with.

alamb commented:
Sorry for the delay @james727  -- this type of PR takes a non trivial amount of time for me to review ;) I need to ensure I have focus time to do so. Going in now

alamb commented on datafusion/src/optimizer/filter_push_down.rs:
```diff
@@ -400,63 +412,68 @@ fn optimize(plan: &LogicalPlan, mut state: State) -> Result<LogicalPlan> {
-                            } else if col == r {
-                                join_cols_to_replace.insert(col, l);
-                                break;
+            if *join_type == JoinType::Inner {
```
this makes sense  (to only create synthetic predicates for inner joins for now).

I will try and capture some of the potential follow on improvements in follow on tickets

alamb reviewed (approved):
Thanks @james727  I went through this carefully and the code now makes a lot of sense to me. 

I will try and file some fun follow on tickets about more join tricks that can be done.

alamb commented:
Epic work @james727 -- thanks again. Kudos to @xudong963  for  the help reviewing

alamb commented:
Follow on https://github.com/apache/arrow-datafusion/issues/1670

xudong963 commented:
Thanks @james727 !
